File: 26_dp_ii/06_word_break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        return self.helper(s, set(wordDict), 0, {})

    def helper(self, s, words, ind, dp):
        if ind in dp:
            return dp[ind]

        if ind == len(s):
            return True

        for i in range(ind + 1, len(s) + 1):
            if s[ind:i] in words and self.helper(s, words, i, dp):
                dp[ind] = True
                return dp[ind]

        dp[ind] = False
        return dp[ind]

File: 26_dp_ii/06_word_break.go
package main

func wordBreak(s string, words []string) bool {

	wordDict := map[string]bool{}
	for _, word := range words {
		wordDict[word] = true
	}
	dp := map[int]bool{}
	var helper func(ind int) bool

	helper = func(ind int) bool {
		if val, ok := dp[ind]; ok {
			return val
		}
		if ind == len(s) {
			return true
		}
		for i := ind + 1; i <= len(s); i++ {
			_, ok := wordDict[s[ind:i]]
			if ok && helper(i) {
				dp[ind] = true
				return true
			}
		}
		dp[ind] = false
		return false
	}
	return helper(0)
}
